crates/bevy_pbr/src/render_graph/forward_pipeline/forward.frag

```diff
@@ -24,41 +27,170 @@ layout(set = 1, binding = 0) uniform Lights {
+    return perceptualRoughness * perceptualRoughness;
+}
+
+#if defined(TARGET_MOBILE)
+// min roughness such that (MIN_PERCEPTUAL_ROUGHNESS^4) > 0 in fp16 (i.e.
+// 2^(-14/4), rounded up)
+#define MIN_PERCEPTUAL_ROUGHNESS 0.089
+#define MIN_ROUGHNESS 0.007921
+#else
+#define MIN_PERCEPTUAL_ROUGHNESS 0.045
+#define MIN_ROUGHNESS 0.002025
+#endif
+
+#endif
 
 void main() {
     vec4 output_color = Albedo;
-# ifdef STANDARDMATERIAL_ALBEDO_TEXTURE
-    output_color *= texture(
-        sampler2D(StandardMaterial_albedo_texture, StandardMaterial_albedo_texture_sampler),
-        v_Uv);
-# endif
-
-# ifndef STANDARDMATERIAL_UNLIT
-    vec3 normal = normalize(v_Normal);
+#ifdef STANDARDMATERIAL_ALBEDO_TEXTURE
+    output_color *= texture(sampler2D(StandardMaterial_albedo_texture,
+                                      StandardMaterial_albedo_texture_sampler),
+                            v_Uv);
+#endif
+
+#ifndef STANDARDMATERIAL_UNLIT
+    float perceptual_roughness = pbr.x;
+    perceptual_roughness =
+        clamp(perceptual_roughness, MIN_PERCEPTUAL_ROUGHNESS, 1.0);
+    float roughness = perceptualRoughnessToRoughness(perceptual_roughness);
+
+    float metallic = pbr.y;
+    vec3 N = normalize(v_Normal);
+    vec3 V = normalize(CameraPos.xyz - v_Position.xyz);
+
+    vec3 F0 = vec3(0.04);
+    F0 = mix(F0, output_color.rgb, metallic);
+
+    vec3 diffuseColor = computeDiffuseColor(output_color.rgb, metallic);
+
     // accumulate color
-    vec3 color = AmbientColor;
-    for (int i=0; i<int(NumLights.x) && i<MAX_LIGHTS; ++i) {
+    vec3 light_accum = vec3(0.0);
+    for (int i = 0; i < int(NumLights.x) && i < MAX_LIGHTS; ++i) {
         Light light = SceneLights[i];
-        // compute Lambertian diffuse term
-        vec3 light_dir = normalize(light.pos.xyz - v_Position);
-        float diffuse = max(0.0, dot(normal, light_dir));
-        // add light contribution
-        color += diffuse * light.color.xyz;
+
+        vec3 lightDir = light.pos.xyz - v_Position.xyz;
+        vec3 L = normalize(lightDir);
+
+        float rangeAttenuation =
+            getDistanceAttenuation(lightDir, light.attenuation);
+
+        vec3 H = normalize(L + V);
+
+        float NdotL = clampNoV(dot(N, L));
+        float NdotV = clampNoV(dot(N, V));
+        float NoL = saturate(NdotL);
+        float NoH = saturate(dot(N, H));
+        float LoH = saturate(dot(L, H));
+
+        vec3 specular = isotropicLobe(F0, roughness, H, NdotV, NoL, NoH, LoH);
+        vec3 diffuse = diffuseColor * Fd_Burley(roughness, NdotV, NoL, LoH);
+
+        light_accum +=
+            ((diffuse + specular) * light.color.xyz) * (light.color.w * NdotL);
     }
 
     // average the lights so that we will never get something with > 1.0
-    color /= max(float(NumLights.x), 1.0);
+    // color /= max(float(NumLights.x), 1.0);
+
+    output_color.xyz = light_accum;
 
-    output_color.xyz *= color;
-# endif
+    // Gamma correction.
```

Comment:
Wouldn't it be more proper to call this Tonemapping?

Reply:
Yeah. I replaced simple reinhard with reindhard luminance.